packages/core/src/new-api/internal/deployment-loader/file-deployment-loader.ts

```diff
@@ -0,0 +1,131 @@
+import fs from "fs-extra";
+import path from "path";
+
+import { Artifact, BuildInfo } from "../../types/artifact";
+import { DeploymentLoader } from "../../types/deployment-loader";
+import { Journal } from "../../types/journal";
+import { FileJournal } from "../journal/file-journal";
+import { MemoryJournal } from "../journal/memory-journal";
+import { assertIgnitionInvariant } from "../utils/assertions";
+
+export class FileDeploymentLoader implements DeploymentLoader {
+  public journal: Journal;
+
+  private _paths: {
+    deploymentDir: string;
+    artifactsDir: string;
+    buildInfoDir: string;
+    journalPath: string;
+    deployedAddressesPath: string;
+  } | null = null;
+
+  constructor(private readonly _deploymentDirPath: string) {
+    this.journal = new MemoryJournal();
+  }
+
+  public async initialize(): Promise<void> {
+    const artifactsDir = path.join(this._deploymentDirPath, "artifacts");
+    const buildInfoDir = path.join(this._deploymentDirPath, "build-info");
+    const journalPath = path.join(this._deploymentDirPath, "journal.jsonl");
+    const deployedAddressesPath = path.join(
+      this._deploymentDirPath,
+      "deployed_addresses.json"
+    );
+
+    this._paths = {
+      deploymentDir: this._deploymentDirPath,
+      artifactsDir,
+      buildInfoDir,
+      journalPath,
+      deployedAddressesPath,
+    };
+
+    await fs.ensureDir(this._paths.deploymentDir);
+    await fs.ensureDir(this._paths.artifactsDir);
+    await fs.ensureDir(this._paths.buildInfoDir);
+    await fs.ensureFile(this._paths.journalPath);
+    await fs.ensureFile(this._paths.deployedAddressesPath);
+
+    this.journal = new FileJournal(journalPath);
+  }
+
+  public async storeArtifact(
+    futureId: string,
+    artifact: Artifact
+  ): Promise<string> {
+    assertIgnitionInvariant(
+      this._paths !== null,
+      "Cannot record deploy address until initialized"
+    );
```

Comment:
I was tempted to have a deployment loader and a deployment loader factory, with the factory doing the initialization and returning a fully setup deployment loader; but I think I can live with these checks over the added complexity.

Reply:
oh, i just proposed an alternative in a comment above